qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -371,47 +496,50 @@ def _add_result_data(self, result: Result) -> None:
-
-        Args:
-            data: Data to be added.
-        """
-        self._data.append(data)
+        with self._data.lock:
+            # Lock data while adding all result data
+            for i, _ in enumerate(result.results):
+                data = result.data(i)
+                data["job_id"] = result.job_id
+                if "counts" in data:
+                    # Format to Counts object rather than hex dict
+                    data["counts"] = result.get_counts(i)
+                expr_result = result.results[i]
+                if hasattr(expr_result, "header") and hasattr(expr_result.header, "metadata"):
+                    data["metadata"] = expr_result.header.metadata
+                data["shots"] = expr_result.shots
+                data["meas_level"] = expr_result.meas_level
+                if hasattr(expr_result, "meas_return"):
+                    data["meas_return"] = expr_result.meas_return
+                self._data.append(data)
 
     def _retrieve_data(self):
         """Retrieve job data if missing experiment data."""
+        if self._data or not self._backend:
+            return
         # Get job results if missing experiment data.
-        if (not self._data) and self._backend:
-            with self._jobs.lock:
-                for jid, job in self._jobs.items():
-                    if job is None:
-                        try:
-                            job = self._backend.retrieve_job(jid)
-                            self._jobs[jid] = job
-                        except Exception:  # pylint: disable=broad-except
-                            LOG.warning(
-                                "Unable to retrive data from job %s on backend %s",
-                                jid,
-                                self._backend,
-                            )
-                    if job is not None:
-                        self._add_result_data(job.result())
+        retrieved_jobs = {}
+        for jid, job in self._jobs.items():
+            if job is None:
+                try:
+                    LOG.debug("Retrieving job from backend %s [Job ID: %s]", self._backend, jid)
+                    job = self._backend.retrieve_job(jid)
+                    retrieved_jobs[jid] = job
+                except Exception:  # pylint: disable=broad-except
+                    LOG.warning(
+                        "Unable to retrive data from job on backend %s [Job ID: %s]",
+                        self._backend,
+                        jid,
+                    )
+        # Add retrieved job objects to stored jobs and extract data
+        for jid, job in retrieved_jobs.items():
+            self._jobs[jid] = job
+            if job.status() in JOB_FINAL_STATES:
+                # Add job results synchronously
+                self._add_job_data(job)
+            else:
+                # Add job results asynchronously
+                self._add_job_future(job)
```

Comment:
Changing to calling `add_jobs` would raise a bunch of warnings about existing job ids at the moment. It could be good to think about changing later so that backend+service can be extracted from job for expdata saved/loaded to JSON, since the backend/provider cant be serialized/deserialized, but that might be better to treat in the serialization PR.